plugins/org.eclipse.n4js.ide/src/org/eclipse/n4js/ide/server/build/N4JSWorkspaceBuilder.java

```diff
@@ -0,0 +1,150 @@
+import org.eclipse.n4js.utils.ProjectDiscoveryHelper;
+import org.eclipse.xtext.resource.IResourceDescription.Event;
+import org.eclipse.xtext.util.CancelIndicator;
+
+import com.google.common.collect.Iterables;
+import com.google.inject.Inject;
+
+/**
+ * Temporary solution until GH-1810 is merged.
+ */
+public class N4JSWorkspaceBuilder extends XWorkspaceBuilder {
+	private static final Logger LOG = LogManager.getLogger(N4JSWorkspaceBuilder.class);
+
+	@Inject
+	private IN4JSCore n4jsCore;
+
+	@Inject
+	private TestCatalogSupplier testCatalogSupplier;
+
+	@Inject
+	private XWorkspaceManager workspaceManager;
+
+	@Inject
+	private NodeModulesDiscoveryHelper nodeModulesDiscoveryHelper;
+
+	@Inject
+	private ProjectDiscoveryHelper projectDiscoveryHelper;
+
+	@Override
+	public BuildTask createIncrementalBuildTask(List<URI> dirtyFiles, List<URI> deletedFiles) {
+		BuildTask buildTask = super.createIncrementalBuildTask(dirtyFiles, deletedFiles);
+
+		BuildTask wrapperBuildTask = new BuildTask() {
+			@Override
+			public Event build(CancelIndicator cancelIndicator) {
+				Event returnedEvent = buildTask.build(cancelIndicator);
+				writeTestCatalogs();
+				return returnedEvent;
+			}
+		};
+		return wrapperBuildTask;
+	}
+
+	@Override
+	public BuildTask createInitialBuildTask() {
+		BuildTask buildTask = super.createInitialBuildTask();
+
+		BuildTask wrapperBuildTask = new BuildTask() {
+			@Override
+			public Event build(CancelIndicator cancelIndicator) {
+				Event returnedEvent = buildTask.build(cancelIndicator);
+				writeTestCatalogs();
+				return returnedEvent;
+			}
+		};
+		return wrapperBuildTask;
+	}
+
+	private void writeTestCatalogs() {
+		Set<IN4JSProject> projects = new HashSet<>();
+		Iterables.addAll(projects, n4jsCore.findAllProjects());
+
+		while (!projects.isEmpty()) {
+			IN4JSProject nextProject = projects.iterator().next();
+			Set<IN4JSProject> catalogsCovered = writeTestCatalog(nextProject);
+			projects.removeAll(catalogsCovered);
+		}
+	}
+
+	private Set<IN4JSProject> writeTestCatalog(IN4JSProject project) {
+		Path projectLocation = project.getLocation().toFileURI().toFile().toPath();
+		NodeModulesFolder nmf = nodeModulesDiscoveryHelper.getNodeModulesFolder(projectLocation);
+		File nmfFile = nmf.isYarnWorkspace() ? nmf.workspaceNodeModulesFolder : nmf.localNodeModulesFolder;
+		File projectRoot = nmfFile.getParentFile();
+		File testCatalog = new File(projectRoot, N4JSGlobals.TEST_CATALOG);
+
+		Set<IN4JSProject> catalogsCovered = new HashSet<>();
+		catalogsCovered.add(project);
+		if (nmf.isYarnWorkspace()) {
+			IN4JSProject yarnProject = n4jsCore.findProject(URI.createFileURI(projectRoot.getPath())).orNull();
+			if (yarnProject != null) {
+				catalogsCovered.add(yarnProject);
+				Set<Path> yarnSubProjectPaths = new HashSet<>();
+				projectDiscoveryHelper.collectYarnWorkspaceProjects(projectRoot.toPath(), new HashMap<>(),
+						yarnSubProjectPaths);
```

Comment:
as discussed: unfortunate that this reads the workspace config from disk (but this code is only temporary anyway).